pallets/gear/src/internal.rs

```diff
@@ -903,22 +906,40 @@ where
         );
     }
 
-    pub(crate) fn inheritor_for(program_id: ProgramId) -> ProgramId {
+    pub(crate) fn inheritor_for(
+        program_id: ProgramId,
+        max_depth: NonZeroUsize,
+    ) -> Option<(ProgramId, BTreeSet<ProgramId>)> {
+        let get_inheritor =
+            |id| Self::exit_inheritor_of(id).or_else(|| Self::termination_inheritor_of(id));
+
+        let max_depth = max_depth.get();
+
         let mut inheritor = program_id;
+        let mut holders: BTreeSet<_> = [program_id].into();
+
+        loop {
+            let next_inheritor = get_inheritor(inheritor)?;
 
-        let mut visited_ids: BTreeSet<_> = [program_id].into();
+            inheritor = next_inheritor;
 
-        while let Some(id) =
-            Self::exit_inheritor_of(inheritor).or_else(|| Self::termination_inheritor_of(inheritor))
-        {
-            if !visited_ids.insert(id) {
+            // don't insert user or active program
+            // because it's the final inheritor we already return
+            if get_inheritor(next_inheritor).is_none() {
                 break;
             }
 
-            inheritor = id
+            if holders.len() == max_depth {
+                break;
+            }
+
+            if !holders.insert(next_inheritor) {
+                log::warn!("Cyclic inheritor detected for {next_inheritor}");
```

Comment:
It is normal situation for gear runtime - so, why do we need to `warn` about this?
```suggestion
                log::debug!("Cyclic inheritor detected for {next_inheritor}");
```